File: src/reduct/api/bucket_api.cc
// Copyright 2022 ReductStore
// This Source Code Form is subject to the terms of the Mozilla Public
//    License, v. 2.0. If a copy of the MPL was not distributed with this
//    file, You can obtain one at https://mozilla.org/MPL/2.0/.

#include "reduct/api/bucket_api.h"

namespace reduct::api {

using core::Error;
using core::Result;
using proto::api::BucketSettings;
using proto::api::FullBucketInfo;
using storage::IStorage;

Result<HttpRequestReceiver> BucketApi::CreateBucket(IStorage* storage, std::string_view name) {
  return ReceiveJson<BucketSettings>([storage, bucket = std::string(name)](BucketSettings&& settings) -> Error {
    return storage->CreateBucket(bucket, settings);
  });
}

Result<HttpRequestReceiver> BucketApi::GetBucket(const IStorage* storage, std::string_view name) {
  auto [bucket_ptr, err] = storage->GetBucket(std::string(name));
  if (err) {
    return err;
  }

  auto bucket = bucket_ptr.lock();
  FullBucketInfo info;

  info.mutable_info()->CopyFrom(bucket->GetInfo());
  info.mutable_settings()->CopyFrom(bucket->GetSettings());
  for (const auto& entry : bucket->GetEntryList()) {
    *info.add_entries() = entry;
  }

  return SendJson<FullBucketInfo>({std::move(info), Error::kOk});
}

core::Result<HttpRequestReceiver> BucketApi::HeadBucket(const storage::IStorage* storage, std::string_view name) {
  auto [bucket_ptr, err] = storage->GetBucket(std::string(name));
  if (err) {
    return err;
  }

  return DefaultReceiver();
}

core::Result<HttpRequestReceiver> BucketApi::UpdateBucket(const storage::IStorage* storage, std::string_view name) {
  auto [bucket_ptr, err] = storage->GetBucket(std::string(name));
  if (err) {
    return err;
  }

  return ReceiveJson<BucketSettings>(
      [bucket = bucket_ptr.lock()](auto settings) { return bucket->SetSettings(settings); });
}
core::Result<HttpRequestReceiver> BucketApi::RemoveBucket(storage::IStorage* storage,
                                                          reduct::rust_part::TokenRepository& repo,
                                                          std::string_view name) {
  if (auto err = storage->RemoveBucket(std::string(name))) {
    return err;
  }

  // Remove bucket from all tokens
  auto err = reduct::rust_part::token_repo_remove_bucket_from_tokens(repo, name.data());
  if (err->status() != Error::kOk.code) {
    return Error(err->status(), err->message().data());
  }

  return DefaultReceiver();
}

}  // namespace reduct::api

File: reductstore/src/http_frontend/bucket_api.rs
// Copyright 2023 ReductStore
// This Source Code Form is subject to the terms of the Mozilla Public
//    License, v. 2.0. If a copy of the MPL was not distributed with this
//    file, You can obtain one at https://mozilla.org/MPL/2.0/.

mod create;
mod get;
mod head;
mod remove;
mod update;

use std::sync::Arc;

use axum::extract::FromRequest;
use axum::headers::HeaderMapExt;
use axum::http::{Request, StatusCode};
use axum::response::{IntoResponse, Response};
use axum::routing::{delete, get, head, post, put};
use axum::{async_trait, headers};
use bytes::Bytes;
use hyper::HeaderMap;

use crate::http_frontend::bucket_api::create::create_bucket;
use crate::http_frontend::bucket_api::get::get_bucket;
use crate::http_frontend::bucket_api::head::head_bucket;
use crate::http_frontend::bucket_api::remove::remove_bucket;
use crate::http_frontend::bucket_api::update::update_bucket;

use crate::http_frontend::{HttpError, HttpServerState};
use reduct_base::msg::bucket_api::{BucketSettings, FullBucketInfo};
use reduct_macros::IntoResponse;
//
// BucketSettings wrapper
//
#[derive(IntoResponse)]
pub struct BucketSettingsAxum(BucketSettings);

impl Default for BucketSettingsAxum {
    fn default() -> Self {
        Self(BucketSettings::default())
    }
}

impl From<BucketSettings> for BucketSettingsAxum {
    fn from(settings: BucketSettings) -> Self {
        Self(settings)
    }
}

impl Into<BucketSettings> for BucketSettingsAxum {
    fn into(self) -> BucketSettings {
        self.0
    }
}

#[derive(IntoResponse)]
pub struct FullBucketInfoAxum(FullBucketInfo);

impl From<FullBucketInfo> for FullBucketInfoAxum {
    fn from(info: FullBucketInfo) -> Self {
        Self(info)
    }
}

impl Into<FullBucketInfo> for FullBucketInfoAxum {
    fn into(self) -> FullBucketInfo {
        self.0
    }
}

#[async_trait]
impl<S, B> FromRequest<S, B> for BucketSettingsAxum
where
    Bytes: FromRequest<S, B>,
    B: Send + 'static,
    S: Send + Sync,
{
    type Rejection = Response;

    async fn from_request(req: Request<B>, state: &S) -> Result<Self, Self::Rejection> {
        let body = Bytes::from_request(req, state)
            .await
            .map_err(IntoResponse::into_response)?;

        let settings: BucketSettings =
            serde_json::from_slice(&body).map_err(|e| HttpError::from(e).into_response())?;
        Ok(Self(settings))
    }
}

pub fn create_bucket_api_routes() -> axum::Router<Arc<HttpServerState>> {
    axum::Router::new()
        .route("/:bucket_name", get(get_bucket))
        .route("/:bucket_name", head(head_bucket))
        .route("/:bucket_name", post(create_bucket))
        .route("/:bucket_name", put(update_bucket))
        .route("/:bucket_name", delete(remove_bucket))
}

#[cfg(test)]
mod tests {
    use super::*;

    use axum::http::Method;
    use hyper::Body;

    #[tokio::test]
    async fn test_bucket_settings_quota_parsing() {
        let req = Request::builder()
            .method(Method::POST)
            .uri("/b/bucket-1")
            .header("Content-Type", "application/json")
            .body(Body::from(r#"{"quota_type": 1}"#))
            .unwrap();
        let resp = BucketSettingsAxum::from_request(req, &())
            .await
            .err()
            .unwrap();
        assert_eq!(resp.status(), 422);
        assert_eq!(
            resp.headers()
                .get("x-reduct-error")
                .unwrap()
                .to_str()
                .unwrap(),
            "Invalid JSON: expected value at line 1 column 16"
        );
    }
}
